File: problems/leetcode/0400/410/410.cpp
// 410. Split Array Largest Sum
// https://leetcode.com/problems/split-array-largest-sum/description/

using namespace std;

#include <vector>

class Solution {
public:
  int splitArray(vector<int> &nums, int k) {
    int n = nums.size();
    vector<int> sums(n, 0);

    vector<int> prefix(n + 1, 0);
    for (int i = 1; i <= n; i++) {
      prefix[i] = prefix[i - 1] + nums[i - 1];
      sums[i - 1] = prefix[i];
    }

    for (int j = 2; j <= k; j++) {
      vector<int> newSums(sums);

      for (int i = 0; i < n; i++) {
        int sum = INT_MAX;

        for (int k = 0; k < i; k++) {
          sum = min(sum, max(sums[k], (prefix[i + 1] - prefix[k + 1])));
        }

        newSums[i] = sum;
      }

      sums = newSums;
    }

    return sums[n - 1];
  }
};

// 1. Naive
// O((n + 1) C k * n)

// 2. DP
// dp[i][j]: i번째까지 j개의 split으로 쪼갰을 때의 min of max of sums
// dp[i][j] = min_(k = 0 to i)(max(dp[k][j - 1], sum btw k + 1 to i))

File: problems/leetcode/0400/410/410.py
# 410. Split Array Largest Sum
# https://leetcode.com/problems/split-array-largest-sum/description

# 1. Naive Solution
# 특정 subarray의 합을 구한 뒤, 남은 subarray에서 k - 1개만큼의 subarray를 고려하는 것을 이용해서 DP 사용 가능

# 2. Parametric Search
# 구간 내에서 특정 합 이하로 subarray들을 만들었을 때 개수가 k개 미만인지를 반복해서 체크

from typing import *

class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        left, right = max(nums), sum(nums)

        while left < right:
            maxSum = (left + right) // 2

            if self.getSubarrayCountWithMaxSum(nums, maxSum) <= k:
                right = maxSum
            else:
                left = maxSum + 1

        return left

    def getSubarrayCountWithMaxSum(self, nums: List[int], maxSum: int) -> int:
        currentSum = 0
        subarrays = 1

        for num in nums:
            if currentSum + num <= maxSum:
                currentSum += num
            else:
                currentSum = num
                subarrays += 1

        return subarrays
